src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py

```diff
@@ -0,0 +1,119 @@
+"""Extract, clean, and normalize the EPA-EIA crosswalk.
+
+This module defines functions that read the raw EPA-EIA crosswalk file, clean
+up the column names, and separate it into three distinctive normalize tables
+for integration in the database.
+
+The crosswalk file was a joint effort on behalf on EPA and EIA and is published on the
+EPA's github account at www.github.com/USEPA".
+"""
+import logging
+
+import pandas as pd
+
+import pudl
+from pudl.metadata.fields import apply_pudl_dtypes
+from pudl.workspace.datastore import Datastore
+
+logger = logging.getLogger(__name__)
+
+
+def extract(ds: Datastore) -> pd.DataFrame:
+    """Extract the EPACEMS-EIA Crosswalk from the Datastore."""
+    with ds.get_zipfile_resource(
+        "epacems_unitid_eia_plant_crosswalk",
+        name="epacems_unitid_eia_plant_crosswalk.zip",
+    ).open("camd-eia-crosswalk-master/epa_eia_crosswalk.csv") as f:
+        return pd.read_csv(f)
+
+
+def transform(
+    epa_eia_crosswalk: pd.DataFrame, generators_entity_eia: pd.DataFrame
+) -> dict[str, pd.DataFrame]:
+    """Clean up the EPACEMS-EIA Crosswalk file and split it into normalized tables.
+
+    Args:
+        epa_eia_crosswalk: The result of running this module's extract() function.
+        generators_entity_eia: The generators_entity_eia table.
+
+    Returns:
+        A dictionary of three normalized DataFrames comprised of the data
+        in the original crosswalk file. EPA plant id to EPA unit id; EPA plant
+        id to EIA plant id; and EIA plant id to EIA generator id to EPA unit
+        id. Includes no nan values.
+    """
+    logger.info("Cleaning up the epacems-eia crosswalk")
+
+    column_rename = {
+        "camd_unit_id": "unit_id_epa",
+        "camd_plant_id": "plant_id_epa",
+        "eia_plant_name": "plant_name_eia",
+        "eia_plant_id": "plant_id_eia",
+        "eia_generator_id": "generator_id",
+    }
+
+    # Basic column rename, selection, and dtype alignment.
+    crosswalk_clean = (
+        epa_eia_crosswalk.pipe(pudl.helpers.simplify_columns)
+        .rename(columns=column_rename)
+        .filter(list(column_rename.values()))
+        .pipe(apply_pudl_dtypes, "eia")
+    )
+
+    # The crosswalk is a static file: there is no year field. The plant_id_eia and
+    # generator_id fields, however, are foreign keys from an annualized table. If the
+    # fast ETL is run (on one year of data) the test will break because the crosswalk
+    # tables with plant_id_eia and generator_id contain values from various years. To
+    # keep the crosswalk in alignment with the available eia data, we'll restrict it
+    # based on the generator entity table which has plant id and generator id.
+    crosswalk_clean = crosswalk_clean.merge(
+        generators_entity_eia[["plant_id_eia", "generator_id"]],
+        on=["plant_id_eia", "generator_id"],
+        how="inner",
+    )
```

Comment:
This seems like a fine workaround for the fast tests, but as you note it's guaranteed to "work" even when it shouldn't, and I think we should be using all the IDs when we're doing the real full ETL or the full integration tests with all years of data.

The only two cases that we're testing and ensuring work are processing all of the years of data, or a single (most recent) year of data. We could apply this workaround conditionally, depending on what years of EIA 860 data are being processed.  If it's all available years, then we don't use the workaround, and we'll see errors that come up as a result of real mismatches between EIA 860 and the crosswalk.

In the main `etl` module you should have access to the `Eia860Settings` at some point, and can pull the years out of that, and compare them to the EIA-860 DataSource metadata, which will tell you what the full set of potentially processable years of data are.

Also, do we know if EPA is committed to updating the crosswalk year over year? As soon as we bring in the new EPA CEMS it seems like there will be new generators that aren't mapped...

Reply:
This is a good idea re: using the workaround only for the fast test.

Regarding the update frequency of the EPA crosswalk...I don't know. It seems like something that will be slow to update given the effort that went into creating it. But it's a git repo so I can open a discussion / create an issue asking this! (or see if there already is one).